File: objc/Project8/Project8/GameScene.m
//
//  GameScene.m
//  Project8
//
//  Created by Jinwoo Kim on 2/12/21.
//

#import "GameScene.h"
#import "ButtonNode.h"
#import "PlayScene.h"

@interface GameScene ()
@property ButtonNode *startNewGame;
@property ButtonNode *howToPlay;
@end

@implementation GameScene

- (void)setup {
    self.startNewGame = [ButtonNode spriteNodeWithImageNamed:@"startNewGameLo"];
    self.howToPlay = [ButtonNode spriteNodeWithImageNamed:@"howToPlayLo"];
}

- (void)didMoveToView:(SKView *)view {
    [self setup];
    [super didMoveToView:view];
    self.anchorPoint = CGPointMake(0.5, 0.5);
    
    SKSpriteNode *background = [SKSpriteNode spriteNodeWithImageNamed:@"background"];
    background.zPosition = -1;
    background.blendMode = SKBlendModeReplace;
    [self addChild:background];
    
    SKSpriteNode *title = [SKSpriteNode spriteNodeWithImageNamed:@"blastazapTitle"];
    title.position = CGPointMake(0, 200);
    [self addChild:title];
    
    SKSpriteNode *strap = [SKSpriteNode spriteNodeWithImageNamed:@"strap"];
    strap.position = CGPointMake(0, 50);
    [self addChild:strap];
    
    //
    
    [self.startNewGame setFocusedImageNamed:@"startNewGameHi"];
    self.startNewGame.position = CGPointMake(0, -100);
    [self addChild:self.startNewGame];
    
    [self.howToPlay setFocusedImageNamed:@"howToPlayHi"];
    self.howToPlay.position = CGPointMake(0, -250);
    [self addChild:self.howToPlay];
}

- (void)didUpdateFocusInContext:(UIFocusUpdateContext *)context withAnimationCoordinator:(UIFocusAnimationCoordinator *)coordinator {
    ButtonNode *previous = (ButtonNode *)context.previouslyFocusedItem;
    ButtonNode *next = (ButtonNode *)context.nextFocusedItem;
    
    if (previous) [previous didLoseFocus];
    if (next) [next didGainFocus];
}

- (void)pressesBegan:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event {
    id selected = UIScreen.mainScreen.focusedItem;
    UIPress *press = [[presses allObjects] firstObject];
    if ((selected == nil) || (press == nil)) return;
    
    if (press.type == UIPressTypeSelect) {
        if (selected == self.startNewGame) {
            PlayScene *game = [PlayScene sceneWithSize:self.size];
            SKTransition *transition = [SKTransition doorsOpenVerticalWithDuration:1];
            [self.view presentScene:game transition:transition];
        }
    }
}

@end

File: swift/Project8/Project8/GameScene.swift
//
//  GameScene.swift
//  Project8
//
//  Created by Jinwoo Kim on 2/12/21.
//

import SpriteKit
import GameplayKit

class GameScene: SKScene {
    let startNewGame = ButtonNode(imageNamed: "startNewGameLo")
    let howToPlay = ButtonNode(imageNamed: "howToPlayLo")
    
    override func didMove(to view: SKView) {
        super.didMove(to: view)
        anchorPoint = CGPoint(x: 0.5, y: 0.5)
        
        let background = SKSpriteNode(imageNamed: "background")
        background.zPosition = -1
        background.blendMode = .replace
        addChild(background)
        
        let title = SKSpriteNode(imageNamed: "blastazapTitle")
        title.position = CGPoint(x: 0, y: 200)
        addChild(title)
        
        let strap = SKSpriteNode(imageNamed: "strap")
        strap.position = CGPoint(x: 0, y: 50)
        addChild(strap)
        
        //
        
        startNewGame.setFocusedImage(named: "startNewGameHi")
        startNewGame.position = CGPoint(x: 0, y: -100)
        addChild(startNewGame)
        
        howToPlay.setFocusedImage(named: "howToPlayHi")
        howToPlay.position = CGPoint(x: 0, y: -250)
        addChild(howToPlay)
    }
    
    override func didUpdateFocus(in context: UIFocusUpdateContext, with coordinator: UIFocusAnimationCoordinator) {
        if let previous = context.previouslyFocusedItem as? ButtonNode {
            previous.didLoseFocus()
        }
        
        if let next = context.nextFocusedItem as? ButtonNode {
            next.didGainFocus()
        }
    }
    
    override func pressesBegan(_ presses: Set<UIPress>, with event: UIPressesEvent?) {
        guard let selected = UIScreen.main.focusedItem else { return }
        guard let press = presses.first else { return }
        
        if press.type == .select {
            if selected === startNewGame {
                let game = PlayScene(size: size)
                let transition = SKTransition.doorsOpenVertical(withDuration: 1)
                view?.presentScene(game, transition: transition)
            }
        }
    }
}
